airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py

```diff
@@ -0,0 +1,122 @@
+    """
+
+    request_body_json: Optional[RequestInput] = None
+    request_headers: Optional[RequestInput] = None
+    request_parameters: Optional[RequestInput] = None
+    request_body_data: Optional[RequestInput] = None
+
+    def __post_init__(self, parameters: Mapping[str, Any]) -> None:
+
+        self.request_options_provider = InterpolatedRequestOptionsProvider(
+            request_body_data=self.request_body_data,
+            request_body_json=self.request_body_json,
+            request_headers=self.request_headers,
+            request_parameters=self.request_parameters,
+            config=self.config,
+            parameters=parameters or {},
+        )
+        super().__post_init__(parameters)
+
+    def get_url_base(self) -> str:
+        self.get_data_center_location()
+        return super().get_url_base()
+
+    def get_data_center_location(self):
+        if not self.config.get("data_center"):
+            if isinstance(self.authenticator, BasicHttpAuthenticator):
+                data_center = self.config["credentials"]["apikey"].split("-").pop()
+            else:
+                data_center = self.get_oauth_data_center(self.config["credentials"]["access_token"])
+            self.config["data_center"] = data_center
+
+    @staticmethod
+    def get_oauth_data_center(access_token: str) -> str:
+        """
+        Every Mailchimp API request must be sent to a specific data center.
+        The data center is already embedded in API keys, but not OAuth access tokens.
+        This method retrieves the data center for OAuth credentials.
+        """
+        response = requests.get("https://login.mailchimp.com/oauth2/metadata", headers={"Authorization": "OAuth {}".format(access_token)})
+
+        # Requests to this endpoint will return a 200 status code even if the access token is invalid.
+        error = response.json().get("error")
+        if error == "invalid_token":
+            raise AirbyteTracedException(
+                failure_type=FailureType.config_error,
+                internal_message=error,
+                message="The access token you provided was invalid. Please check your credentials and try again.",
+            )
+        return response.json()["dc"]
+
+
+class MailChimpRecordFilter(RecordFilter):
+    """
+    Filter applied on a list of Records.
+    """
+
+    parameters: InitVar[Mapping[str, Any]]
+
+    def __post_init__(self, parameters: Mapping[str, Any]) -> None:
+        self.parameters = parameters
+
+    def filter_records(
+        self,
+        records: List[Mapping[str, Any]],
+        stream_state: StreamState,
+        stream_slice: Optional[StreamSlice] = None,
+        next_page_token: Optional[Mapping[str, Any]] = None,
+    ) -> List[Mapping[str, Any]]:
+        current_state = [x for x in stream_state.get("states", []) if x["partition"]["id"] == stream_slice.partition["id"]]
+        # TODO: REF what to do if no start_date mentioned (see manifest)
```

Comment:
Why is this TODO needed? Shouldn't we just accept everything if there are no states (which is the logic you implemented here)?

Reply:
agreed